File: lib/tv/widgets/tv_keyboard.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import '../utils/tv_utils.dart';
import '../utils/tv_keyboard_focus_manager.dart';

typedef OnKeyboardInput = void Function(String text);

class TvKeyboard extends StatefulWidget {
  final OnKeyboardInput onInput;
  final VoidCallback onSubmit;
  final String initialText;
  final TvKeyboardFocusManager? focusManager;

  const TvKeyboard({
    super.key,
    required this.onInput,
    required this.onSubmit,
    this.initialText = '',
    this.focusManager,
  });

  @override
  State<TvKeyboard> createState() => _TvKeyboardState();
}

class _TvKeyboardState extends State<TvKeyboard> {
  late List<List<String>> _keyboardLayout;
  int _selectedRow = 0;
  int _selectedCol = 0;
  String _input = '';
  bool _capsLock = false;
  bool _isSymbols = false;
  late FocusNode _keyboardFocusNode;

  @override
  void initState() {
    super.initState();
    _input = widget.initialText;
    _initializeKeyboard();
    _keyboardFocusNode = FocusNode();

    // Activate keyboard focus on init
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _keyboardFocusNode.requestFocus();
      widget.focusManager?.activateKeyboard();
    });
  }

  @override
  void dispose() {
    _keyboardFocusNode.dispose();
    widget.focusManager?.deactivateKeyboard();
    super.dispose();
  }

  void _initializeKeyboard() {
    if (_isSymbols) {
      _keyboardLayout = [
        ['!', '@', '#', '\$', '%', '^', '&', '*', '(', ')'],
        ['-', '_', '=', '+', '[', ']', '{', '}', '|', '\\'],
        [';', ':', '\'', '"', '<', '>', ',', '.', '?', '/'],
        ['SPACE', 'BACKSPACE', 'CLEAR', 'ABC', 'DONE'],
      ];
    } else {
      _keyboardLayout = [
        ['Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P'],
        ['A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L'],
        ['Z', 'X', 'C', 'V', 'B', 'N', 'M'],
        ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0'],
        ['SPACE', 'BACKSPACE', 'CLEAR', 'SYM', 'CAPS', 'DONE'],
      ];
    }
  }

  void _handleKeyEvent(RawKeyEvent event) {
    // Only handle arrow keys if keyboard is focused
    if (event.isKeyPressed(LogicalKeyboardKey.arrowUp)) {
      setState(() {
        _selectedRow = (_selectedRow - 1).clamp(0, _keyboardLayout.length - 1);
        _selectedCol = _selectedCol.clamp(
          0,
          _keyboardLayout[_selectedRow].length - 1,
        );
      });
    } else if (event.isKeyPressed(LogicalKeyboardKey.arrowDown)) {
      setState(() {
        _selectedRow = (_selectedRow + 1).clamp(0, _keyboardLayout.length - 1);
        _selectedCol = _selectedCol.clamp(
          0,
          _keyboardLayout[_selectedRow].length - 1,
        );
      });
    } else if (event.isKeyPressed(LogicalKeyboardKey.arrowLeft)) {
      setState(() {
        _selectedCol = (_selectedCol - 1).clamp(
          0,
          _keyboardLayout[_selectedRow].length - 1,
        );
      });
    } else if (event.isKeyPressed(LogicalKeyboardKey.arrowRight)) {
      setState(() {
        _selectedCol = (_selectedCol + 1).clamp(
          0,
          _keyboardLayout[_selectedRow].length - 1,
        );
      });
    } else if (event.isKeyPressed(LogicalKeyboardKey.select) ||
        event.isKeyPressed(LogicalKeyboardKey.enter)) {
      _pressKey(_keyboardLayout[_selectedRow][_selectedCol]);
    } else if (event.isKeyPressed(LogicalKeyboardKey.escape) ||
        event.isKeyPressed(LogicalKeyboardKey.goBack)) {
      // If back is pressed on keyboard, go back to content
      widget.focusManager?.focusOnContent();
    }
  }

  void _pressKey(String key) {
    setState(() {
      if (key == 'SPACE') {
        _input += ' ';
      } else if (key == 'BACKSPACE') {
        if (_input.isNotEmpty) {
          _input = _input.substring(0, _input.length - 1);
        }
      } else if (key == 'CLEAR') {
        _input = '';
      } else if (key == 'CAPS') {
        _capsLock = !_capsLock;
      } else if (key == 'SYM') {
        _isSymbols = !_isSymbols;
        _selectedRow = 0;
        _selectedCol = 0;
        _initializeKeyboard();
        return;
      } else if (key == 'ABC') {
        _isSymbols = false;
        _selectedRow = 0;
        _selectedCol = 0;
        _initializeKeyboard();
        return;
      } else if (key == 'DONE') {
        widget.onSubmit();
        return;
      } else {
        if (!_isSymbols && _capsLock) {
          _input += key;
        } else if (!_isSymbols && !_capsLock) {
          _input += key.toLowerCase();
        } else {
          _input += key;
        }
      }
      widget.onInput(_input);
    });
  }

  @override
  Widget build(BuildContext context) {
    return RawKeyboardListener(
      focusNode: _keyboardFocusNode,
      onKey: _handleKeyEvent,
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          // Input display
          Container(
            padding: EdgeInsets.all(TvUtils.responsivePadding(16, context)),
            decoration: BoxDecoration(
              color: const Color(0xFF2A2A2A),
              borderRadius: BorderRadius.circular(
                TvUtils.responsivePadding(12, context),
              ),
              border: Border.all(color: Colors.grey, width: 2),
            ),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Expanded(
                  child: Text(
                    _input.isEmpty ? 'Start typing...' : _input,
                    style: TextStyle(
                      color: _input.isEmpty ? Colors.grey : Colors.white,
                      fontSize: TvUtils.responsiveFontSize(
                        20,
                        context,
                        maxSize: 36,
                      ),
                    ),
                    maxLines: 2,
                    overflow: TextOverflow.ellipsis,
                  ),
                ),
                if (_input.isNotEmpty)
                  GestureDetector(
                    onTap: () {
                      setState(() => _input = '');
                      widget.onInput('');
                    },
                    child: Icon(
                      Icons.clear,
                      color: Colors.red,
                      size: TvUtils.responsiveFontSize(
                        24,
                        context,
                        maxSize: 40,
                      ),
                    ),
                  ),
              ],
            ),
          ),
          SizedBox(height: TvUtils.responsivePadding(24, context)),

          // Keyboard
          Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              for (int row = 0; row < _keyboardLayout.length; row++)
                Padding(
                  padding: EdgeInsets.only(
                    bottom: row < _keyboardLayout.length - 1
                        ? TvUtils.responsivePadding(12, context)
                        : 0,
                  ),
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    children: [
                      for (
                        int col = 0;
                        col < _keyboardLayout[row].length;
                        col++
                      )
                        Expanded(
                          child: Padding(
                            padding: EdgeInsets.symmetric(
                              horizontal: TvUtils.responsivePadding(4, context),
                            ),
                            child: _buildKey(
                              _keyboardLayout[row][col],
                              row,
                              col,
                            ),
                          ),
                        ),
                    ],
                  ),
                ),

              // Status row
              SizedBox(height: TvUtils.responsivePadding(16, context)),
              Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  if (!_isSymbols) ...[
                    _buildStatus('CAPS', _capsLock),
                    SizedBox(width: TvUtils.responsivePadding(16, context)),
                  ],
                  _buildStatus(_isSymbols ? 'ABC' : 'SYM', _isSymbols),
                ],
              ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _buildKey(String key, int row, int col) {
    final isFocused = _selectedRow == row && _selectedCol == col;
    final isSpecial = [
      'SPACE',
      'BACKSPACE',
      'CLEAR',
      'SYM',
      'ABC',
      'CAPS',
      'DONE',
    ].contains(key);

    return GestureDetector(
      onTap: () => _pressKey(key),
      child: Container(
        height: TvUtils.responsiveButtonHeight(context) * 0.5,
        decoration: BoxDecoration(
          color: isFocused
              ? const Color(0xFFE50914)
              : (isSpecial ? Colors.grey[700] : Colors.grey[800]),
          border: Border.all(
            color: isFocused ? Colors.white : Colors.transparent,
            width: isFocused ? 2 : 1,
          ),
          borderRadius: BorderRadius.circular(
            TvUtils.responsivePadding(6, context),
          ),
        ),
        child: Center(
          child: Text(
            key,
            style: TextStyle(
              color: isFocused ? Colors.white : Colors.white70,
              fontSize: TvUtils.responsiveFontSize(12, context, maxSize: 20),
              fontWeight: isFocused ? FontWeight.bold : FontWeight.normal,
            ),
            textAlign: TextAlign.center,
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
        ),
      ),
    );
  }

  Widget _buildStatus(String label, bool isActive) {
    return Container(
      padding: EdgeInsets.symmetric(
        horizontal: TvUtils.responsivePadding(12, context),
        vertical: TvUtils.responsivePadding(6, context),
      ),
      decoration: BoxDecoration(
        color: isActive ? const Color(0xFFE50914) : Colors.grey[700],
        borderRadius: BorderRadius.circular(
          TvUtils.responsivePadding(6, context),
        ),
      ),
      child: Text(
        label,
        style: TextStyle(
          color: Colors.white,
          fontSize: TvUtils.responsiveFontSize(12, context, maxSize: 18),
          fontWeight: FontWeight.bold,
        ),
      ),
    );
  }
}

File: android/tvapp/src/main/java/com/maxstream/app/ui/components/TvKeyboard.kt
package com.maxstream.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.focusable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.key.Key
import androidx.compose.ui.input.key.KeyEventType
import androidx.compose.ui.input.key.key
import androidx.compose.ui.input.key.onKeyEvent
import androidx.compose.ui.input.key.type
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.maxstream.app.ui.tv.TvKeyboardFocusManager

@Composable
fun TvKeyboard(
    onInput: (String) -> Unit,
    onSubmit: () -> Unit,
    initialText: String = "",
    focusManager: TvKeyboardFocusManager? = null,
    onMoveRight: (() -> Unit)? = null,
    onMoveLeft: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    var keyboardLayout by remember { mutableStateOf(emptyList<List<String>>()) }
    var selectedRow by remember { mutableStateOf(0) }
    var selectedCol by remember { mutableStateOf(0) }
    var input by remember { mutableStateOf(initialText) }
    var capsLock by remember { mutableStateOf(false) }
    var isSymbols by remember { mutableStateOf(false) }
    val keyboardFocusRequester = remember { FocusRequester() }

    fun initializeKeyboard() {
        keyboardLayout = if (isSymbols) {
            listOf(
                listOf("!", "@", "#", "\$", "%", "^", "&", "*", "(", ")"),
                listOf("-", "_", "=", "+", "[", "]", "{", "}", "|", "\\"),
                listOf(";", ":", "'", "\"", "<", ">", ",", ".", "?", "/"),
                listOf("SPACE", "BACKSPACE", "CLEAR", "ABC", "DONE"),
            )
        } else {
            listOf(
                listOf("Q", "W", "E", "R", "T", "Y", "U", "I", "O", "P"),
                listOf("A", "S", "D", "F", "G", "H", "J", "K", "L"),
                listOf("Z", "X", "C", "V", "B", "N", "M"),
                listOf("1", "2", "3", "4", "5", "6", "7", "8", "9", "0"),
                listOf("SPACE", "BACKSPACE", "CLEAR", "SYM", "CAPS", "DONE"),
            )
        }
    }

    fun pressKey(key: String) {
        when (key) {
            "SPACE" -> input += " "
            "BACKSPACE" -> {
                if (input.isNotEmpty()) {
                    input = input.dropLast(1)
                }
            }
            "CLEAR" -> input = ""
            "CAPS" -> capsLock = !capsLock
            "SYM" -> {
                isSymbols = true
                selectedRow = 0
                selectedCol = 0
                initializeKeyboard()
                return
            }
            "ABC" -> {
                isSymbols = false
                selectedRow = 0
                selectedCol = 0
                initializeKeyboard()
                return
            }
            "DONE" -> {
                onSubmit()
                return
            }
            else -> {
                input += if (!isSymbols && capsLock) {
                    key
                } else if (!isSymbols && !capsLock) {
                    key.lowercase()
                } else {
                    key
                }
            }
        }
        onInput(input)
    }

    LaunchedEffect(Unit) {
        initializeKeyboard()
        keyboardFocusRequester.requestFocus()
        focusManager?.activateKeyboard()
    }

    Column(
        modifier = modifier
            .focusRequester(keyboardFocusRequester)
            .focusable()
            .onKeyEvent { event ->
                if (event.type == KeyEventType.KeyDown) {
                    when (event.key) {
                        Key.DirectionUp -> {
                            selectedRow = (selectedRow - 1).coerceAtLeast(0)
                            selectedCol = selectedCol.coerceAtMost(keyboardLayout.getOrNull(selectedRow)?.lastIndex ?: 0)
                            true
                        }
                        Key.DirectionDown -> {
                            selectedRow = (selectedRow + 1).coerceAtMost(keyboardLayout.lastIndex)
                            selectedCol = selectedCol.coerceAtMost(keyboardLayout.getOrNull(selectedRow)?.lastIndex ?: 0)
                            true
                        }
                        Key.DirectionLeft -> {
                            if (selectedCol == 0) {
                                onMoveLeft?.invoke()
                                true
                            } else {
                                selectedCol = (selectedCol - 1).coerceAtLeast(0)
                                true
                            }
                        }
                        Key.DirectionRight -> {
                            val currentRow = keyboardLayout.getOrNull(selectedRow)
                            if (currentRow != null && selectedCol == currentRow.lastIndex) {
                                onMoveRight?.invoke()
                                true
                            } else {
                                selectedCol = (selectedCol + 1).coerceAtMost(currentRow?.lastIndex ?: 0)
                                true
                            }
                        }
                        Key.Enter -> {
                            val key = keyboardLayout.getOrNull(selectedRow)?.getOrNull(selectedCol)
                            if (key != null) {
                                pressKey(key)
                            }
                            true
                        }
                        Key.Escape -> {
                            onMoveLeft?.invoke()
                            true
                        }
                        else -> false
                    }
                } else false
            }
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .background(Color(0xFF2A2A2A), RoundedCornerShape(12.dp))
                .border(2.dp, Color.Gray, RoundedCornerShape(12.dp))
                .padding(16.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                androidx.compose.material3.Text(
                    text = if (input.isEmpty()) "Start typing..." else input,
                    color = if (input.isEmpty()) Color.Gray else Color.White,
                    fontSize = 20.sp,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f)
                )
                if (input.isNotEmpty()) {
                    androidx.compose.material3.Text(
                        text = "Clear",
                        color = Color.Red,
                        fontSize = 20.sp,
                        modifier = Modifier
                            .clickable {
                                input = ""
                                onInput("")
                            }
                            .padding(start = 8.dp)
                    )
                }
            }
        }

        Spacer(modifier = Modifier.height(24.dp))

        Column(
            modifier = Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            keyboardLayout.forEachIndexed { rowIndex, row ->
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(bottom = if (rowIndex < keyboardLayout.lastIndex) 12.dp else 0.dp),
                    horizontalArrangement = Arrangement.SpaceEvenly
                ) {
                    row.forEachIndexed { colIndex, key ->
                        val isFocused = selectedRow == rowIndex && selectedCol == colIndex
                        val isSpecial = key in listOf("SPACE", "BACKSPACE", "CLEAR", "SYM", "ABC", "CAPS", "DONE")

                        Box(
                            modifier = Modifier
                                .weight(1f)
                                .padding(horizontal = 4.dp)
                                .height(56.dp)
                                .clip(RoundedCornerShape(6.dp))
                                .background(
                                    when {
                                        isFocused -> Color(0xFFE50914)
                                        isSpecial -> Color(0xFF666666)
                                        else -> Color(0xFF4D4D4D)
                                    }
                                )
                                .border(
                                    width = if (isFocused) 2.dp else 1.dp,
                                    color = if (isFocused) Color.White else Color.Transparent,
                                    shape = RoundedCornerShape(6.dp)
                                )
                                .clickable { pressKey(key) },
                            contentAlignment = Alignment.Center
                        ) {
                            androidx.compose.material3.Text(
                                text = key,
                                color = if (isFocused) Color.White else Color.White.copy(alpha = 0.7f),
                                fontSize = 12.sp,
                                fontWeight = if (isFocused) FontWeight.Bold else FontWeight.Normal,
                                textAlign = TextAlign.Center,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis
                            )
                        }
                    }
                }
            }

            Spacer(modifier = Modifier.height(16.dp))

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.Center
            ) {
                if (!isSymbols) {
                    TvKeyboardStatusBadge(
                        label = "CAPS",
                        isActive = capsLock,
                        modifier = Modifier.padding(end = 16.dp)
                    )
                }
                TvKeyboardStatusBadge(
                    label = if (isSymbols) "ABC" else "SYM",
                    isActive = isSymbols
                )
            }
        }
    }
}

@Composable
private fun TvKeyboardStatusBadge(
    label: String,
    isActive: Boolean,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .padding(horizontal = 12.dp, vertical = 6.dp)
            .clip(RoundedCornerShape(6.dp))
            .background(if (isActive) Color(0xFFE50914) else Color(0xFF666666))
            .padding(horizontal = 12.dp, vertical = 6.dp),
        contentAlignment = Alignment.Center
    ) {
        androidx.compose.material3.Text(
            text = label,
            color = Color.White,
            fontSize = 12.sp,
            fontWeight = FontWeight.Bold
        )
    }
}
